src/net/atlas/download.rs

```diff
@@ -0,0 +1,1025 @@
+            }
+        };
+    }
+
+    pub fn bump_retry_count(&mut self) {
+        self.retry_count += 1;
+    }
+
+    pub fn has_fully_succeed(&self) -> bool {
+        self.attachments_instances_count == 0
+    }
+
+    pub fn get_missing_pages_for_contract_id(
+        &self,
+        contract_id: &QualifiedContractIdentifier,
+    ) -> Vec<u32> {
+        let mut pages_indexes = HashSet::new();
+        if let Some(missing_attachments) = self.attachments_instances.get(&contract_id) {
+            for ((page_index, _), _) in missing_attachments.iter() {
+                pages_indexes.insert(*page_index);
+            }
+        }
+        pages_indexes.into_iter().collect()
+    }
+
+    pub fn resolve_attachment(&mut self, content_hash: &Hash160) {
+        for missing_attachments in self.attachments_instances.values_mut() {
+            let mut keys = vec![];
+            for (k, hash) in missing_attachments.iter() {
+                if hash == content_hash {
+                    keys.push(k.clone());
+                }
+            }
+            for key in keys {
+                missing_attachments.remove(&key);
+                self.attachments_instances_count =
+                    self.attachments_instances_count.saturating_sub(1);
+            }
+        }
+    }
+
+    pub fn get_stacks_block_id(&self) -> StacksBlockId {
+        StacksBlockHeader::make_index_block_hash(&self.consensus_hash, &self.block_header_hash)
+    }
+}
+
+impl Ord for AttachmentsBatch {
+    fn cmp(&self, other: &AttachmentsBatch) -> Ordering {
+        other
+            .retry_count
+            .cmp(&self.retry_count)
+            .then_with(|| {
+                self.attachments_instances_count
+                    .cmp(&other.attachments_instances_count)
+            })
+            .then_with(|| other.block_height.cmp(&self.block_height))
+    }
+}
+
+impl PartialOrd for AttachmentsBatch {
+    fn partial_cmp(&self, other: &AttachmentsBatch) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
+pub struct ReliabilityReport {
+    pub total_requests_sent: u32,
+    pub total_requests_success: u32,
+}
+
+impl ReliabilityReport {
+    pub fn bump_successful_requests(&mut self) {
+        self.total_requests_sent += 1;
+        self.total_requests_success += 1;
+    }
+
+    pub fn bump_failed_requests(&mut self) {
+        self.total_requests_sent += 1;
+    }
+}
+
+impl ReliabilityReport {
+    pub fn new(total_requests_sent: u32, total_requests_success: u32) -> ReliabilityReport {
+        ReliabilityReport {
+            total_requests_sent,
+            total_requests_success,
+        }
+    }
+
+    pub fn empty() -> ReliabilityReport {
+        ReliabilityReport {
+            total_requests_sent: 0,
+            total_requests_success: 0,
+        }
+    }
+
+    pub fn score(&self) -> u32 {
+        match self.total_requests_sent {
+            0 => 0 as u32,
```

Comment:
What does a score of `0` mean here?  It can't mean "100% unreliable," because a node with a reliability report of 0 may simply be a node that was never contacted.

Reply:
I'd suggest making the `0` case a random value.  The `0` case simply means that you don't know how reliable the node is.  Also, you'll want some randomness in your reliability ranking, because otherwise a network attacker can eclipse you by making you believe that all but the attacker-controlled nodes are completely unreliable.  Adding some randomness forces the attacker to _continuously_ block your access to honest nodes, whereas using a deterministic ranking allows the attacker to optimize the attack by only blocking honest nodes until you no longer believe that can be relied upon.